### ISSUE TITLE
Intermittent failure to retrieve advertised content

### ISSUE THREAD

**Issue by gammazero** [filecoin-project/indexer-reference-provider#49]
Sometimes after publishing an advertisement, when the indexer asks the reference provider for the content of that advertisement, the reference provider cannot find the content.  It seems like there is some delay before it is available from the linksystem cache.

This can be forced to work by re-importing the car file and making the reference provider re-publish its advertisement, which triggers the indexer to ask for content again… and it usually works after one or two retries.  Of course, this is not acceptable. 

This problem appears to be on the reference provider side, because it appears that the indexer is asking for exactly the same content, in exactly the same way/order for both the success and failure case.

**Comment by gammazero** [comment]
I can put a retry loop in the `StorageReadOpener` function returned by `mkLinkSystem()`, [here](https://github.com/filecoin-project/indexer-reference-provider/blob/main/engine/linksystem.go#L28), and eventually, it will always work after a seemingly random number of retries.  The retry loop MUST include `generateChunks` or it never works.

The retries are not affected by time.  In other words, having a delay between retries does not change the behavior.   To make it work, simply retrying regardless of any delay is what is needed.

**Comment by gammazero** [comment]
This will be fixed when #37 is resolved.

**Comment by gammazero** [comment]
What is happening is that sometimes, but not always, the CAR iterator, used by the car_suppier is returning multihashes in a different order for the same car file.

I put a debug print [here](https://github.com/filecoin-project/indexer-reference-provider/blob/main/index_mh_iter.go#L32)
```go
fmt.Println("mh from CAR:", mh.B58String())
```
Two separate reads of the same CAR file generated a slightly different order of the last handful of multihashes.  The different order causes the chunk CID to be different, resulting in different runs of CAR chunking having different IPLD links.  Here is what the difference looks like; only showing the last 7 multihashes since everything else (1000+ multihashes) is the same:
Iteration 1:
```
mh from CAR: 2DrjgbMSZJkSSM3pNf9ZzS5NCWSTsYHLn8b4A7QjRJEy54hqjn
mh from CAR: 15vj7gjjx5PBq2VFzahy
mh from CAR: 122uRfmEvYNy1J7SXwR
mh from CAR: 13aabCA6uJRydQkM
mh from CAR: 13aabCA6uJS8Vz4K
mh from CAR: 1DUhBwk6syC9SnaVX
mh from CAR: 1DUhBwk6syCVEL1wo
```
Iteration 2:
```
mh from CAR: 2DrjgbMSZJkSSM3pNf9ZzS5NCWSTsYHLn8b4A7QjRJEy54hqjn
mh from CAR: 13aabCA6uJRydQkM
mh from CAR: 13aabCA6uJS8Vz4K
mh from CAR: 1DUhBwk6syC9SnaVX
mh from CAR: 1DUhBwk6syCVEL1wo
mh from CAR: 15vj7gjjx5PBq2VFzahy
mh from CAR: 122uRfmEvYNy1J7SXwR
```
The solution is for the CAR iteration results to be sorted.

**Comment by willscott** [comment]
https://github.com/ipld/go-car/pull/258

**Comment by gammazero** [comment]
Fixed by [ipld/go-car#258](https://github.com/ipld/go-car/pull/258) and  https://github.com/filecoin-project/indexer-reference-provider/commits/main
